### DISCUSSION TITLE
Images Randomly Swapped in Production

### DISCUSSION THREAD

**Discussion by jonathanhnr** [vercel/next.js#73062] [Help]
### Summary

I'm facing a curious problem with my Next.js application in production. At times, many images on the site appear swapped. When I directly access the URL of the changed image, it loads correctly. This problem does not occur in a development environment, only in production.

### Additional information

```js
Here are some more details about the issue:
1. Current Behavior:
    - On some pages, several images appear swapped for others.
    - When you inspect the element and follow the image URL, it opens correctly in the browser.

2. Production Environment:
    - Next.js is being hosted: AWS
    - Next.js version: 14.2.4

3. Solution Attempts:
    - I've already tried clearing the cache and rebuilding the project, but the problem persists.
    - I checked the encoding of the URLs and everything seems to be correct.
    - I also checked the configuration in the `next.config.js` file, and the domains are configured correctly.


Any ideas what could be causing this problem or any steps I can take to debug this in production?```


### Example

_No response_

**Comment by muezz** [comment]
Did you ever find a solution for this? This only happens in prod for me so it is very hard for me to test if any solutions might work.

**Reply by NicolasPineau** [reply]
Having a similar issue on next 12.2.5. Image are swapped but curiously each of them is "shifted" of exactly one place... the issue is'nt related to a misuse of "key" attribute of components. Up to know I was'nt able to set any reproducible context... does anyone has a clue ?

**Comment by sefin23** [comment]
This is a known issue with Next.js Image Optimization and how the image cache works in production. Here are the most likely causes and how to fix them:

## Root Cause

Next.js uses a server-side image optimization cache (stored in `.next/cache/images`). In production (especially on AWS with multiple instances/containers), **each instance maintains its own local cache**, and images can get **mismatched cache keys** — particularly when:

- You have **multiple server instances** (horizontal scaling) and requests are load-balanced across them
- The image cache gets a **hash/key collision** between different image URLs
- You're using a **CDN in front of Next.js** that caches responses incorrectly

## Solutions

### 1. Use a remote image cache (recommended for multi-instance AWS)

Configure a custom cache handler so all instances share the same cache:

```js
// next.config.js
module.exports = {
  cacheHandler: require.resolve('./cache-handler.js'),
  cacheMaxMemorySize: 0, // disable in-memory caching
}
```

Or use [`@neshca/cache-handler`](https://github.com/caching-tools/next-shared-cache) with Redis.

### 2. Set `minimumCacheTTL` and proper cache headers

Make sure your images have unique, stable URLs and configure proper cache TTLs:

```js
// next.config.js
module.exports = {
  images: {
    minimumCacheTTL: 60,
    deviceSizes: [640, 750, 828, 1080, 1200],
  },
}
```

### 3. Check for missing or duplicate `key` props

Even though NicolasPineau ruled this out, it's worth double-checking — if you're rendering a **list of images without stable keys**, React can reuse DOM nodes incorrectly:

```jsx
// Bad
images.map((img) => <Image src={img.url} ... />)

// Good
images.map((img) => <Image key={img.id} src={img.url} ... />)
```

### 4. Disable Next.js image optimization and use your own CDN

If you're self-hosting on AWS, consider offloading image optimization to CloudFront or a dedicated service:

```js
// next.config.js
module.exports = {
  images: {
    loader: 'custom',
    loaderFile: './my-loader.js',
  },
}
```

### 5. Check your load balancer / sticky sessions

If you have multiple EC2 instances or ECS tasks, make sure your ALB (Application Load Balancer) uses **sticky sessions** or, better yet, move to a shared cache solution.

## Quick Debug Steps

1. Check if the issue stops when you scale down to **a single instance** — this confirms a multi-instance cache problem
2. Look at the `x-nextjs-cache` response header — if it says `HIT` when it should be fresh, your cache is the culprit
3. Try setting `Cache-Control: no-store` temporarily on image responses to rule out CDN caching

Hope this helps narrow it down!
